### PR TITLE
Add LinkManager plugin

### PR DESCRIPTION
**Pull request by dougmacknz**
Includes the complete plugin for managing links, including the popover and the modal. To see it action: https://dev.cultureamp.design/KDS-436/add-link-functionality/?path=/story/components-rich-text-editor--default

### CODE REVIEW

**Comment by mcwinter07** [comment]
I think I found the route of the error in the build, looks like we need to add a types.d.ts to our root to allow for our naming conventions then add that to our tsconfig files :) 

```
declare module "*.scss" {
  const classes: { [key: string]: string }
  export default classes
}

declare module "*.icon.svg" {
  const content: {
    id: string
    viewBox: string
  }
  export default content
}
```

This seems to be enough to resolve those angry error (at least locally)

File: src/plugins/LinkManager/types.d.ts [hunk 1]
```diff
@@ -0,0 +1,29 @@
+import { EditorState, Transaction } from "prosemirror-state"
+import { EditorView } from "prosemirror-view"
+
+/*
+ * Dispatch receives a transaction from ProseMirror to effect changes to the
+ * EditorState in order
+ */
+export type Dispatch = (tx: Transaction) => void
+
+export type Command = (
```
Comment (mcwinter07) [inline src/plugins/LinkManager/types.d.ts]: Not sure if it'll help here but `prosemirror-commands` also exports the command type

File: src/plugins/LinkManager/createReactTooltipWrapper.tsx [hunk 2]
```diff
@@ -0,0 +1,66 @@
+import Nanobus from "nanobus"
```
Comment (mcwinter07) [inline src/plugins/LinkManager/createReactTooltipWrapper.tsx:2]: Can't say I've used nanobus before but I think I grasp the general straws of it. Did have a question more out of curiosity that feedback (so no rush on responding) I'm a little hazy on the scope of the listener (in this case "update") - ie: is it tied to the specific component or could you emit an event anywhere in the DOM if the Wrapper was mounted or is it tied to an emitter that has to be passed in

**Review by mcwinter07 — DISMISSED** [review]
Looks good as far as the eye can see ⛵ Happy to give this a tick before I jet off to approve changes up to this point 👍

File: src/plugins/LinkManager/types.d.ts (hunk 1 above)
Comment (dougmacknz) [reply src/plugins/LinkManager/types.d.ts]: Nice spot

File: src/plugins/LinkManager/components/LinkModal/LinkModal.tsx [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+interface LinkModalProps {
+  onSubmit: (href: string) => void
+  onDismiss: () => void
+  onAfterLeave: () => void
+  isOpen: boolean
+  defaultHref?: string
+}
+
+export const LinkModal: React.VFC<LinkModalProps> = props => {
+  const { onSubmit, onDismiss, onAfterLeave, isOpen, defaultHref } = props
```
Comment (NatalieLoyzaga) [inline src/plugins/LinkManager/components/LinkModal/LinkModal.tsx:14]: Minor - Could line 14 be de-structured in line 13 replacing props with `({ onSubmit, onDismiss, onAfterLeave, isOpen, defaultHref })`

File: src/plugins/LinkManager/components/LinkModal/LinkModal.tsx [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+    >
+      <TextField
+        id="href"
+        type="text"
+        defaultValue={href ?? ""}
+        labelText="Link URL"
+        inputRef={inputRef}
+        validationMessage={validationError ? "Please enter a URL" : undefined}
+        status={validationError ? "error" : undefined}
+        onChange={e => {
```
Comment (NatalieLoyzaga) [inline src/plugins/LinkManager/components/LinkModal/LinkModal.tsx]: Add type to event `(e:  React.ChangeEvent<HTMLInputElement>)`

File: src/plugins/LinkManager/components/LinkPopover/LinkPopover.tsx [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+
+export interface LinkPopoverProps {
+  href?: string
+  onRemove: () => void
+  onEdit: () => void
+  selectionPosition: SelectionPosition
+}
+
+export const LinkPopover: React.VFC<LinkPopoverProps> = props => {
+  const { href, onRemove, onEdit, selectionPosition } = props
```
Comment (NatalieLoyzaga) [inline src/plugins/LinkManager/components/LinkPopover/LinkPopover.tsx:21]: Minor - Could line 21 be de-structured in line 20 replacing props with ({ href, onRemove, onEdit, selectionPosition })

File: src/plugins/LinkManager/components/LinkModal/LinkModal.tsx [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+  defaultHref?: string
+}
+
+export const LinkModal: React.VFC<LinkModalProps> = props => {
+  const { onSubmit, onDismiss, onAfterLeave, isOpen, defaultHref } = props
+  const [href, setHref] = useState<string>(defaultHref || "")
+  const [validationError, setValidationError] = useState<boolean>(false)
+  const inputRef = useRef<HTMLInputElement>(null)
+
+  const handleSubmit = () => {
```
Comment (NatalieLoyzaga) [inline src/plugins/LinkManager/components/LinkModal/LinkModal.tsx]: Add return type to function - `const handleSubmit = (): void => {`

File: src/plugins/LinkManager/createReactTooltipWrapper.tsx [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+import React, { ElementType, useEffect, useState } from "react"
+import ReactDOM from "react-dom"
+
+export type TooltipComponentType = string | unknown
```
Comment (NatalieLoyzaga) [inline src/plugins/LinkManager/createReactTooltipWrapper.tsx]: Can this be moved to the types file then be imported in?

**Comment by HeartSquared** [comment]
>I think I found the route of the error in the build, looks like we need to add a types.d.ts to our root to allow for our naming conventions then add that to our tsconfig files :)

Just a little FYI that some certain TS config would actually allow for a more accurate TS typing of CSS modules (unsure about the icons). Had it set up once; although I don't remember exactly what it was anymore haha. Would need to look into it.

File: src/plugins/LinkManager/createReactTooltipWrapper.tsx (hunk 7 above)
Comment (HeartSquared) [reply src/plugins/LinkManager/createReactTooltipWrapper.tsx]: `TooltipComponentType` is a very interesting type name 😅 Would ideally also be better if we could find out what `unknown` should be.

File: src/plugins/LinkManager/components/LinkModal/LinkModal.tsx (hunk 3 above)
Comment (dougmacknz) [reply src/plugins/LinkManager/components/LinkModal/LinkModal.tsx:14]: I personally prefer it this way. It's not always useful, but it does allow to work with the props object as a whole if you ever want to. There might be the odd case where I don't want to destructure a prop and instead reference it from the props object.

For example, I might want to reference a prop in a component, but still have it be part of the `...restProps` spread:
```
export const LinkModal: React.VFC<LinkModalProps> = props => {
  const { onSubmit, onDismiss, onAfterLeave, isOpen, defaultHref, ...restProps } = props
  const someConditional = props["aria-labelledby"] != null
}
```

I wouldn't be able to do that with the props destructured in the function declaration. Or even if I just want to console.log all the props for some reason.

**Review by NatalieLoyzaga — APPROVED** [review]
go for ittt

**Comment by dougmacknz** [comment]
Wew

**Pull request merged**